File: python/core/autostart.py
import sys
import os
import plistlib

from core.logger import logger

APP_NAME = "ClaudeHUDMonitor"
RUN_KEY_PATH = r"Software\Microsoft\Windows\CurrentVersion\Run"

def is_autostart_enabled() -> bool:
    if sys.platform == "win32":
        try:
            import winreg
            with winreg.OpenKey(winreg.HKEY_CURRENT_USER, RUN_KEY_PATH, 0, winreg.KEY_READ) as key:
                value, _ = winreg.QueryValueEx(key, APP_NAME)
                return bool(value)
        except Exception:
            return False
    elif sys.platform == "darwin":
        plist_path = os.path.expanduser(f"~/Library/LaunchAgents/com.claudehud.plist")
        return os.path.exists(plist_path)
    return False

def set_autostart(enable: bool) -> bool:
    if sys.platform == "win32":
        try:
            import winreg
            with winreg.OpenKey(winreg.HKEY_CURRENT_USER, RUN_KEY_PATH, 0, winreg.KEY_SET_VALUE) as key:
                if enable:
                    if getattr(sys, 'frozen', False):
                        cmd = f'"{sys.executable}"'
                    else:
                        pythonw = os.path.join(os.path.dirname(sys.executable), "pythonw.exe")
                        runner = pythonw if os.path.exists(pythonw) else sys.executable
                        script = os.path.abspath(sys.argv[0])
                        cmd = f'"{runner}" "{script}"'
                    winreg.SetValueEx(key, APP_NAME, 0, winreg.REG_SZ, cmd)
                else:
                    try:
                        winreg.DeleteValue(key, APP_NAME)
                    except FileNotFoundError:
                        pass
                return True
        except Exception as e:
            logger.error(f"[AutoStart Windows] Error setting autostart: {e}", exc_info=True)
            return False
    elif sys.platform == "darwin":
        try:
            plist_dir = os.path.expanduser("~/Library/LaunchAgents")
            os.makedirs(plist_dir, exist_ok=True)
            plist_path = os.path.join(plist_dir, "com.claudehud.plist")
            if enable:
                arguments = [sys.executable]
                if not getattr(sys, 'frozen', False):
                    arguments.append(os.path.abspath(sys.argv[0]))
                with open(plist_path, "wb") as f:
                    plistlib.dump({
                        "Label": "com.claudehud",
                        "ProgramArguments": arguments,
                        "RunAtLoad": True
                    }, f)
            else:
                if os.path.exists(plist_path):
                    try:
                        import subprocess
                        subprocess.run(["launchctl", "unload", plist_path], capture_output=True)
                    except Exception:
                        pass
                    os.remove(plist_path)
            return True
        except Exception as e:
            logger.error(f"[AutoStart macOS] Error setting autostart: {e}", exc_info=True)
            return False
    return False

File: rust/src/autostart.rs
// src/autostart.rs — Platform autostart management
// Mirrors Python core/autostart.py

#[cfg(any(target_os = "windows", target_os = "macos"))]
use log::error;

#[cfg(target_os = "windows")]
const APP_NAME: &str = "ClaudeHUDMonitor";

/// Check if autostart is enabled for the current platform.
pub fn is_autostart_enabled() -> bool {
    #[cfg(target_os = "windows")]
    return windows_is_enabled();

    #[cfg(target_os = "macos")]
    return macos_plist_path().exists();

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    false
}

/// Enable or disable autostart. Returns true on success.
pub fn set_autostart(_enable: bool) -> bool {
    #[cfg(target_os = "windows")]
    return windows_set(_enable);

    #[cfg(target_os = "macos")]
    return macos_set(_enable);

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    false
}

// ── Windows ───────────────────────────────────────────────────────────────────
#[cfg(target_os = "windows")]
fn windows_is_enabled() -> bool {
    use winreg::enums::*;
    use winreg::RegKey;
    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
    let Ok(key) = hkcu.open_subkey(r"Software\Microsoft\Windows\CurrentVersion\Run") else {
        return false;
    };
    key.get_value::<String, _>(APP_NAME).is_ok()
}

#[cfg(target_os = "windows")]
fn windows_set(enable: bool) -> bool {
    use winreg::enums::*;
    use winreg::RegKey;
    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
    match hkcu.open_subkey_with_flags(
        r"Software\Microsoft\Windows\CurrentVersion\Run",
        KEY_SET_VALUE,
    ) {
        Ok(key) => {
            if enable {
                let exe = std::env::current_exe()
                    .map(|p| format!("\"{}\"", p.display()))
                    .unwrap_or_default();
                key.set_value(APP_NAME, &exe)
                    .map_err(|e| error!("[AutoStart] {e}"))
                    .is_ok()
            } else {
                let _ = key.delete_value(APP_NAME);
                true
            }
        }
        Err(e) => {
            error!("[AutoStart] Cannot open registry key: {e}");
            false
        }
    }
}

// ── macOS ─────────────────────────────────────────────────────────────────────
#[cfg(target_os = "macos")]
fn macos_plist_path() -> std::path::PathBuf {
    let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_owned());
    std::path::PathBuf::from(home)
        .join("Library")
        .join("LaunchAgents")
        .join("com.claudehud.plist")
}

#[cfg(target_os = "macos")]
fn macos_set(enable: bool) -> bool {
    let plist_path = macos_plist_path();
    if enable {
        let exe = std::env::current_exe()
            .map(|p| p.display().to_string())
            .unwrap_or_default();
        let plist_content = format!(
            r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.claudehud</string>
    <key>ProgramArguments</key>
    <array><string>{}</string></array>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>"#,
            exe
        );
        if let Some(dir) = plist_path.parent() {
            let _ = std::fs::create_dir_all(dir);
        }
        std::fs::write(&plist_path, plist_content)
            .map_err(|e| error!("[AutoStart] {e}"))
            .is_ok()
    } else {
        if plist_path.exists() {
            let _ = std::process::Command::new("launchctl")
                .args(["unload", &plist_path.display().to_string()])
                .output();
            std::fs::remove_file(&plist_path)
                .map_err(|e| error!("[AutoStart] {e}"))
                .is_ok()
        } else {
            true
        }
    }
}
